File: sdk/verifiers/ts/tests/aarp-corpus.test.ts
// Copyright 2026 Josh Waldrep
// SPDX-License-Identifier: Apache-2.0

// Corpus-driven conformance test: run the TS AARP verifier over EVERY fixture in
// the shared hostile corpus and assert appraise fixtures emit bytes equal to the
// committed <name>.appraisal.json (trimmed) and exit 0; fatal fixtures exit
// non-zero. This is the per-language arm of the four-language gate.

import { readFileSync, readdirSync } from "node:fs";
import test from "node:test";
import assert from "node:assert/strict";
import { comparableAppraisal } from "../src/aarp/appraise.js";
import {
  comparableChain,
  loadTrustFile,
  unmarshal,
  verify,
  verifyChain,
} from "../src/aarp/index.js";
import type { Envelope } from "../src/aarp/index.js";
import { appraiseWithSVID, loadSVIDFile } from "../src/aarp/svid.js";

// Paths are relative to the test CWD (sdk/verifiers/ts), matching the existing
// conformance-corpus.test.ts convention.
const corpus = "../../conformance/testdata/aarp-corpus";
const trustPath = `${corpus}/trust.json`;

interface ExpectFile {
  fixture_id: string;
  category: string;
  input_format: string;
  verdict: string;
}

// appraiseOne mirrors the CLI single-envelope path and returns either the
// comparable bytes (appraised) or the marker that it was fatal.
function appraiseOne(text: string): { fatal: boolean; bytes?: string } {
  let env: Envelope;
  try {
    env = unmarshal(text);
  } catch {
    return { fatal: true };
  }
  try {
    const opts = loadTrustFile(trustPath);
    const ap = verify(env, opts);
    return { fatal: false, bytes: comparableAppraisal(ap) };
  } catch {
    return { fatal: true };
  }
}

// appraiseChain mirrors the CLI --chain path. A fatal parse of any line, or a
// stream that does not link, is the fatal/non-zero outcome.
function appraiseChain(text: string): { fatal: boolean; bytes?: string } {
  const lines = text.trim().split("\n");
  const envs: Envelope[] = [];
  for (const raw of lines) {
    const line = raw.trim();
    if (line === "") continue;
    let env: Envelope;
    try {
      env = unmarshal(line);
    } catch {
      return { fatal: true };
    }
    envs.push(env);
  }
  const bytes = comparableChain(envs);
  if (!verifyChain(envs)) return { fatal: true };
  return { fatal: false, bytes };
}

for (const category of ["golden", "malicious", "edge", "chain"]) {
  const dir = `${corpus}/${category}`;
  const expectFiles = readdirSync(dir).filter((f) => f.endsWith(".expect.json"));
  for (const expectFile of expectFiles) {
    const base = expectFile.replace(/\.expect\.json$/u, "");
    const expect = JSON.parse(readFileSync(`${dir}/${expectFile}`, "utf8")) as ExpectFile;
    const isChain = expect.input_format === "chain";
    const fixturePath = `${dir}/${base}${isChain ? ".aarp.jsonl" : ".aarp.json"}`;

    test(`corpus ${category}/${base} (${expect.verdict})`, () => {
      const text = readFileSync(fixturePath, "utf8");
      const result = isChain ? appraiseChain(text) : appraiseOne(text);
      if (expect.verdict === "fatal") {
        assert.equal(result.fatal, true, `${base} should be fatal`);
        return;
      }
      assert.equal(result.fatal, false, `${base} should appraise`);
      const want = readFileSync(`${dir}/${base}.appraisal.json`, "utf8").trim();
      assert.equal(result.bytes, want, `${base} comparable bytes mismatch`);
    });
  }
}

// SVID arm: each fixture is appraised with its per-fixture --svid sidecar and
// must byte-match the committed .appraisal.json. A genuine baseline confirms the
// three workload-identity claims; every attack appraises WITHOUT them (no
// inflation). Mirrors the CLI single-envelope + --svid path.
{
  const dir = `${corpus}/svid`;
  const expectFiles = readdirSync(dir).filter((f) => f.endsWith(".expect.json"));
  for (const expectFile of expectFiles) {
    const base = expectFile.replace(/\.expect\.json$/u, "");
    const expect = JSON.parse(readFileSync(`${dir}/${expectFile}`, "utf8")) as ExpectFile;

    test(`corpus svid/${base} (${expect.verdict})`, () => {
      const env = unmarshal(readFileSync(`${dir}/${base}.aarp.json`, "utf8"));
      const opts = loadTrustFile(trustPath);
      const svid = loadSVIDFile(`${dir}/${base}.svid.json`);
      const ap = appraiseWithSVID(env, svid.evidence, opts, svid.opts);
      const bytes = comparableAppraisal(ap);
      const want = readFileSync(`${dir}/${base}.appraisal.json`, "utf8").trim();
      assert.equal(bytes, want, `${base} comparable bytes mismatch`);
    });
  }
}

File: sdk/conformance/aarp_corpus_test.go
// Copyright 2026 Josh Waldrep
// SPDX-License-Identifier: Apache-2.0

package conformance_test

import (
	"bytes"
	"encoding/json"
	"errors"
	"io/fs"
	"os"
	"path/filepath"
	"strings"
	"testing"

	"github.com/luckyPipewrench/pipelock/internal/aarp"
	"github.com/luckyPipewrench/pipelock/internal/svidsidecar"
)

// This file is the GO arm of the four-language AARP conformance gate. It runs the
// Go reference verifier over every fixture in testdata/aarp-corpus and asserts:
//
//   - "appraise" envelope fixtures verify and emit ComparableAppraisal bytes that
//     match the committed <name>.appraisal.json exactly;
//   - "appraise" chain fixtures link and emit matching ComparableChain bytes;
//   - "fatal" fixtures reject (Unmarshal/Verify error, or a broken chain).
//
// The TypeScript, Rust, and Python verifiers are run over the SAME corpus by
// aarp-corpus-gate.sh; this test is the Go member of that set and runs in the
// standard `test` CI job, so a Go-side regression is caught even when the bash
// gate job is skipped by path filters.

type aarpExpect struct {
	FixtureID   string `json:"fixture_id"`
	Category    string `json:"category"`
	InputFormat string `json:"input_format"`
	Verdict     string `json:"verdict"`
}

// aarpCorpusVerifyOptions reconstructs the pinned trust the corpus was generated
// against, from the same deterministic seeds, so the Go arm needs no external
// trust file at run time.
func aarpCorpusVerifyOptions() aarp.VerifyOptions {
	signerPub, _ := keyFromSeed(seedSigner)
	issuerPub, _ := keyFromSeed(seedIssuer)
	return corpusVerifyOptions(signerPub, issuerPub)
}

func TestAARPCorpus(t *testing.T) {
	t.Parallel()
	opts := aarpCorpusVerifyOptions()

	var checked int
	for _, category := range []string{catGolden, catMalicious, catEdge, catChain, catSVID, catKillSuite} {
		dir := filepath.Join(aarpCorpusDir, category)
		entries, err := os.ReadDir(dir)
		if err != nil {
			t.Fatalf("read corpus dir %s: %v", dir, err)
		}
		for _, e := range entries {
			name := e.Name()
			if !strings.HasSuffix(name, ".expect.json") {
				continue
			}
			base := strings.TrimSuffix(name, ".expect.json")
			t.Run(category+"/"+base, func(t *testing.T) {
				t.Parallel()
				runAARPCorpusFixture(t, dir, base, opts)
			})
			checked++
		}
	}
	if checked == 0 {
		t.Fatal("no AARP fixtures found; corpus generation is broken (run -update-aarp)")
	}
}

func runAARPCorpusFixture(t *testing.T, dir, base string, opts aarp.VerifyOptions) {
	t.Helper()
	exp := readAARPExpect(t, filepath.Join(dir, base+".expect.json"))

	if exp.InputFormat == "chain" {
		runAARPChainFixture(t, dir, base, exp)
		return
	}

	// SVID fixtures carry a sidecar and are appraised through AppraiseWithSVID.
	// They are always verdictAppraise; the committed appraisal encodes whether the
	// workload-identity claims attach (no inflation on the hostile fixtures).
	sidecarPath := filepath.Join(dir, base+".svid.json")
	if _, statErr := os.Stat(sidecarPath); statErr == nil {
		runAARPSVIDFixture(t, dir, base, sidecarPath, exp, opts)
		return
	} else if !errors.Is(statErr, fs.ErrNotExist) {
		// A real filesystem error (e.g. permissions) must fail loudly, not be
		// silently treated as "no sidecar" and fall through to the envelope path.
		t.Fatalf("stat sidecar %s: %v", sidecarPath, statErr)
	}

	body := readFixture(t, filepath.Join(dir, base+".aarp.json"))
	env, err := aarp.Unmarshal(body)
	if err != nil {
		if exp.Verdict == verdictFatal {
			return // expected: fatal at parse time
		}
		t.Fatalf("appraise fixture %s failed to unmarshal: %v", base, err)
	}
	ap, err := aarp.Verify(env, opts)
	if err != nil {
		if exp.Verdict == verdictFatal {
			return // expected: fatal at verify time
		}
		t.Fatalf("appraise fixture %s failed to verify: %v", base, err)
	}
	if exp.Verdict == verdictFatal {
		t.Fatalf("fatal fixture %s unexpectedly appraised: %+v", base, ap)
	}

	got, err := aarp.ComparableAppraisal(ap)
	if err != nil {
		t.Fatalf("comparable appraisal %s: %v", base, err)
	}
	want := readFixture(t, filepath.Join(dir, base+".appraisal.json"))
	assertComparableMatch(t, base, got, want)
}

func runAARPChainFixture(t *testing.T, dir, base string, exp aarpExpect) {
	t.Helper()
	body := readFixture(t, filepath.Join(dir, base+".aarp.jsonl"))
	envs, parseErr := parseChainLines(body)
	if parseErr != nil {
		if exp.Verdict == verdictFatal {
			return
		}
		t.Fatalf("chain fixture %s failed to parse: %v", base, parseErr)
	}

	linked := aarp.VerifyChain(envs) == nil
	if exp.Verdict == verdictFatal {
		if linked {
			t.Fatalf("fatal chain fixture %s unexpectedly linked", base)
		}
		return
	}
	if !linked {
		t.Fatalf("appraise chain fixture %s did not link", base)
	}
	got, err := aarp.ComparableChain(envs)
	if err != nil {
		t.Fatalf("comparable chain %s: %v", base, err)
	}
	want := readFixture(t, filepath.Join(dir, base+".appraisal.json"))
	assertComparableMatch(t, base, got, want)
}

// runAARPSVIDFixture appraises an SVID fixture through the Go reference
// AppraiseWithSVID (envelope + sidecar) and asserts its ComparableAppraisal bytes
// match the committed appraisal. The sidecar is loaded from its on-disk wire form
// — the same bytes the four-language gate feeds every verifier via --svid.
func runAARPSVIDFixture(t *testing.T, dir, base, sidecarPath string, exp aarpExpect, opts aarp.VerifyOptions) {
	t.Helper()
	// SVID attacks are never envelope-fatal: a failed/absent binding withholds the
	// workload-identity claims, it never rejects the envelope. So every SVID
	// fixture must be labeled appraise; a fatal label is a corpus authoring bug.
	if exp.Verdict != verdictAppraise {
		t.Fatalf("svid fixture %s has verdict %q; SVID fixtures are always %q (an attack withholds claims, never fails the envelope)", base, exp.Verdict, verdictAppraise)
	}
	sc, err := svidsidecar.Parse(readFixture(t, sidecarPath))
	if err != nil {
		t.Fatalf("parse svid sidecar %s: %v", base, err)
	}
	body := readFixture(t, filepath.Join(dir, base+".aarp.json"))
	got := svidComparableBytes(t, body, sc, opts)
	want := readFixture(t, filepath.Join(dir, base+".appraisal.json"))
	assertComparableMatch(t, base, got, want)
}

// parseChainLines unmarshals one envelope per non-empty line. A parse error on
// any line is the chain-fatal signal (a malformed envelope in the stream).
func parseChainLines(body []byte) ([]aarp.Envelope, error) {
	var envs []aarp.Envelope
	for _, line := range bytes.Split(bytes.TrimSpace(body), []byte("\n")) {
		if len(bytes.TrimSpace(line)) == 0 {
			continue
		}
		e, err := aarp.Unmarshal(line)
		if err != nil {
			return nil, err
		}
		envs = append(envs, e)
	}
	return envs, nil
}

// assertComparableMatch compares got vs the committed appraisal bytes, tolerating
// only a single trailing newline difference (the committed files end in \n).
func assertComparableMatch(t *testing.T, base string, got, want []byte) {
	t.Helper()
	if !bytes.Equal(bytes.TrimRight(got, "\n"), bytes.TrimRight(want, "\n")) {
		t.Errorf("comparable mismatch for %s:\n got: %s\nwant: %s", base,
			bytes.TrimRight(got, "\n"), bytes.TrimRight(want, "\n"))
	}
}

func readAARPExpect(t *testing.T, path string) aarpExpect {
	t.Helper()
	var exp aarpExpect
	unmarshalJSONFile(t, path, &exp)
	return exp
}

func readFixture(t *testing.T, path string) []byte {
	t.Helper()
	data, err := os.ReadFile(filepath.Clean(path))
	if err != nil {
		t.Fatalf("read fixture %s: %v", path, err)
	}
	return data
}

func unmarshalJSONFile(t *testing.T, path string, v any) {
	t.Helper()
	data := readFixture(t, path)
	if err := json.Unmarshal(data, v); err != nil {
		t.Fatalf("unmarshal %s: %v", path, err)
	}
}
